### PR TITLE
[POC] [Not for merge] Recursive rendering as unknown block fallback

### LINKED ISSUE #1099: Recursively render recognised blocks
**Is your feature request related to a problem? Please describe.**
Similar - but different - to https://github.com/BBC-News/simorgh/issues/1100.

A product requirement from Jim:

```gherkin
When I try to render a block I do not recognise
Then I should not throw an error
But I should recursively attempt to render the block's children instead
```

Example:

```jsx
<UnknownFoo>
    <Text>This is some very useful text</Text>
</UnknownFoo>
```

**Describe alternatives you've considered**

We have 3 options realistically:

1. Throw an error and render an [Error Boundary component](https://reactjs.org/blog/2017/07/26/error-handling-in-react-16.html) instead (e.g. "We could not render this block"), so we never see the "This is some very useful text".
2. Throw an error but suppress it silently (so the audience is unaware of the component at all). Again, we would never see the "This is some very useful text".
3. Render "This is some very useful text" as if it had been provided purely as a `<Text>` component. NB, we'd probably want some additional console.log or HTML comment data for debugging so we can dig in and see that we had to bypass an `UnknownFoo` component. The audience would never need to be aware of this.

We'd like to go for option 3.

**Describe the solution you'd like**
Write a test suite that contains the example JSX above, and tests that Simorgh renders only the `Text` portion of the component.

(Naturally, it will fail the test, so we'll need to write the corresponding code as well).

**Testing notes**
Tester will pair with the developer when this is Ready For Test, given some of the complexities and edge cases that will occur during testing.

Dev insight: see "Describe the solution you'd like"

- [x] Initially labelled with ["Refinement needed"](https://github.com/BBC-News/simorgh/labels/Refinement%20Needed)

### PR DESCRIPTION
**Pull request by bcmn**
Resolves #1099 

**Overall change:** Implements a recursive fallback, where any unknown block is still evaluated to see if any of its child blocks could be shown. Includes a collection of mock-ups of how this would appear, & how this may have to alter the codebase.

**Code changes:**

- Implements the recursion behaviour very simply, as a branch of the existing `Blocks` container.
- As commented inline, I'm not sure this would currently be the correct way to implement this recursion, so we would probably need to profile any real implementation of this thoroughly.
- Currently configured to treat lists, videos & images as unrenderable
- Ideally componentsToRender would just become a list of every block-equivalent container
- Adds data containing a list

---

In the case of List blocks, which currently have an implementation in our data source which we don't support, this behaviour works relatively well:
screenshot 2019-02-20 at 15 52 45

Video (with Caption) before:
screenshot 2019-02-20 at 17 09 14

Video (with Caption) after:
screenshot 2019-02-21 at 12 23 50

Videos currently look quite odd, as their Caption would theoretically be rendered as either a floating Caption, or as contextless Text. However, Jim & I understand this will be mitigated significantly very soon, as the `aresMedia` structure will be changed to contain an Optimo-style Image block -- so a failed Video would appear just like an Image (with any caption below it). This would seem to work very well.

---

I also experimented with disabling the Image block, to see another demonstration of how this behaviour might work. In this case, due to the `altText` being a block in the data structure, it would be rendered as `Text`:

Image (without caption) before:
screenshot 2019-02-20 at 18 43 57

Image (without caption) after:
screenshot 2019-02-20 at 18 44 47

This could potentially be limited by wrapping unrecognised content in something that would help contextualise it. A very quick mock-up:
screenshot 2019-02-20 at 19 04 02

screenshot 2019-02-20 at 19 02 46

Jim is not keen on this contextualisation element, & feels altText should never be shown. If we continue down the recursive path, it's not immediately clear to me how this could be achieved -- the simplest way would need `altText` as a type to either be treated as a container (how this would work in normal rendering I don't know), or have a hard-coded branch in the recursion that knows altText is never displayed like this (this feels somewhat out of line with the very "naive" recursion ideal).

---

It is worth noting all of the examples above (assuming the Video image was updated) would require recursion, & would not work with a simple depth limit of 1.

Another key takeaway is that rendering Captions without a Video or Image is currently quite messy to mock up, & it raises issues with our current `Blocks` implementation. Different containers expect different sections of their data structure to be passed to them. I believe Text, Image, & Headline all expect to receive the `blocks.model.blocks` array from their data structure, but containers like `Caption` currently expect to receive their entire `block` structure. To implement this in a truly recursive way, I think we would have to standardise this for all block-equivalent containers. This would also help serve any future "container library" priorities that may arise.

---

Overall, I'm very positive on this approach. There are two key parts of the codebase that would need to be addressed to go further down this path:

- Standardising block-equivalent container props.
- Handling altText in a way that ensures it is never displayed. (A valid solution to this may be to feed back that we don't want them to be blocks)
- We will not implement any contextualisation around blocks that are rendered due to the fallback.

This fallback solution may need buy-in from Optimo/Ares to design their future block structures with this kind of fallback in mind.

---

- [ ] I have assigned myself to this PR and the corresponding issues
- [ ] Tests added for new features
- [ ] Test engineer approval
- [ ] I have followed [the merging checklist](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/MERGE_PROCESS.md) and this is ready to merge.

### CODE REVIEW

File: src/app/containers/Blocks/index.jsx [hunk 1]
```diff
@@ -8,21 +11,34 @@ const BlockString = props => {
...
 const Blocks = ({ blocks, componentsToRender }) =>
-  blocks.map((block, index) => {
+  blocks.map(block => {
     const { type, model } = block;
+    const subblocks = block.model.blocks;
+
+    if (Object.keys(componentsToRender).includes(type)) {
+      const Block = componentsToRender[type] || BlockString;
+
+      if (type === 'caption') {
```
Comment (ChrisBAshton) [inline src/app/containers/Blocks/index.jsx:29]: I see why you want to keep our block passing consistent! Totally want to avoid this conditional 👍

Comment (bcmn) [reply src/app/containers/Blocks/index.jsx:29]: Haha yep. It's a spike 🤫😉

File: src/app/containers/Blocks/index.jsx [hunk 2]
```diff
@@ -8,21 +11,34 @@ const BlockString = props => {
...
+        return <Block key={nanoid()} type={type} block={block} />;
+      }
 
-    const { type: typeOfPreviousBlock } = blocks[index - 1] || {};
```
Comment (bcmn) [inline src/app/containers/Blocks/index.jsx:15]: Also, FWIW, it looks like none of this logic is used any more.

**Comment by bcmn** [comment]
Closing to implement properly in #1352.

**Comment by ChrisBAshton** [comment]
We'll ignore the `altText` issue for now - this may become more of a generic issue of 'leaf nodes' vs 'container nodes' in the future.
